# Introduction: Multiple DataFrames

In order to efficiently store data, we often spread related information across multiple tables.

For instance, imagine that we own an e-commerce business and we want to track the products that have been ordered from our website.

We could have one table with all of the following information:

* order_id
* customer_id
* customer_name
* customer_address
* customer_phone_number
* product_id
* product_description
* product_price
* quantity
* timestamp

However, a lot of this information would be repeated. If the same customer makes multiple orders, that customer's name, address, and phone number will be reported multiple times. If the same product is ordered by multiple customers, then the product price and description will be repeated. This will make our orders table big and unmanageable.

So instead, we can split our data into three tables:


1. orders would contain just the information necessary to describe what was ordered:
   * product_id
   * customer_id
   * quantity
   * timestamp


2. products would contain the information to describe each product:
   * product_id
   * product_description
   * product_price


3. customers would contain the information for each customer:
   * customer_id
   * customer_name
   * customer_address
   * customer_phone_number

In this lesson, we will learn the Pandas commands that help us work with data stored in multiple tables.

In [1]:
import pandas as pd
orders = pd.read_csv('orders.csv')
print(orders,'\n')
products = pd.read_csv('products.csv')
print(products,'\n')
customers = pd.read_csv('customers.csv')
print(customers,'\n')

   order_id  customer_id  product_id  quantity   timestamp
0         1            2           3         1  2017-01-01
1         2            2           2         3  2017-01-01
2         3            3           1         1  2017-01-01
3         4            3           2         2  2017-02-01
4         5            3           3         3  2017-02-01
5         6            1           4         2  2017-03-01
6         7            1           1         1  2017-02-02
7         8            1           4         1  2017-02-02 

   product_id         description  price
0           1      thing-a-ma-jig      5
1           2  whatcha-ma-call-it     10
2           3          doo-hickey      7
3           4               gizmo      3 

   customer_id customer_name        address  phone_number
0            1    John Smith   123 Main St.  212-123-4567
1            2      Jane Doe  456 Park Ave.  949-867-5309
2            3     Joe Schmo   798 Broadway  112-358-1321 



# Inner Merge

Suppose we have the following three tables that describe our eCommerce business:

* orders — a table with information on each transaction:

<table>
    <tr><th>order_id</th><th>customer_id</th><th>product_id</th><th>quantity</th><th>timestamp</th></tr>
    <tr><td>1</td><td>2</td><td>3</td><td>1</td><td>2017-01-01</td></tr>
    <tr><td>2</td><td>2</td><td>2</td><td>3</td><td>2017-01-01</td></tr>
    <tr><td>3</td><td>3</td><td>1</td><td>1</td><td>2017-01-01</td></tr>
    <tr><td>4</td><td>3</td><td>2</td><td>2</td><td>2017-02-01</td></tr>
    <tr><td>5</td><td>3</td><td>3</td><td>3</td><td>2017-02-01</td></tr>
    <tr><td>6</td><td>1</td><td>4</td><td>2</td><td>2017-03-01</td></tr>
    <tr><td>7</td><td>1</td><td>1</td><td>1</td><td>2017-02-02</td></tr>
    <tr><td>8</td><td>1</td><td>4</td><td>1</td><td>2017-02-02</td></tr>
</table>

* products — a table with product IDs, descriptions, and prices:

<table>
    <tr><th>product_id</th><th>description</th><th>price</th></tr>
    <tr><td>1</td><td>thing-a-ma-jig</td><td>5</td></tr>
    <tr><td>2</td><td>whatcha-ma-call-it</td><td>10</td></tr>
    <tr><td>3</td><td>doo-hickey</td><td>7</td></tr>
    <tr><td>4</td><td>gizmo</td><td>3</td></tr>
</table>

* customers — a table with customer names and contact information:

<table>
    <tr><th>customer_id</th><th>customer_name</th><th>address</th><th>phone_number</th></tr>
    <tr><th>1</th><td>John Smith</td><td>123 Main St.</td><td>212-123-4567</td></tr>
    <tr><th>2</th><td>Jane Doe</td><td>456 Park Ave.</td><td>949-867-5309</td></tr>
    <tr><th>3</th><td>Joe Schmo</td><td>798 Broadway</td><td>112-358-1321</td></tr>
</table>

If we just look at the orders table, we can't really tell what's happened in each order. However, if we refer to the other tables, we can get a more complete picture.

Let's examine the order with an order_id of 1. It was purchased by Customer 2. To find out the customer's name, we look at the customers table and look for the item with a customer_id value of 2. We can see that Customer 2's name is Jane Doe and that she lives at 456 Park Ave.

Doing this kind of matching is called merging two DataFrames.

The .merge method looks for columns that are common between two DataFrames and then looks for rows where those column's values are the same. It then combines the matching rows into a single row in a new table.

We can call the pd.merge method with two tables like this:

`new_df = pd.merge(orders, customers)`

This will match up all of the customer information to the orders that each customer made.


You are an analyst Cool T-Shirts Inc. You are going to help them analyze some of their sales data.

There are two DataFrames defined in the file script.py:

* sales contains the monthly revenue for Cool T-Shirts Inc. It has two columns: month and revenue.
* targets contains the goals for monthly revenue for each month. It has two columns: month and target.

In [7]:
sales = pd.read_csv('sales.csv')
print(sales,'\n')
targets = pd.read_csv('targets.csv')
print(targets)

      month  revenue
0   January      300
1  February      290
2     March      310
3     April      325
4       May      475
5      June      495 

      month  target
0   January     310
1  February     270
2     March     300
3     April     350
4       May     475
5      June     500


Create a new DataFrame sales_vs_targets which contains the merge of sales and targets.

In [9]:
sales_vs_targets = pd.merge(sales, targets)
print(sales_vs_targets)

      month  revenue  target
0   January      300     310
1  February      290     270
2     March      310     300
3     April      325     350
4       May      475     475
5      June      495     500


Cool T-Shirts Inc. wants to know the months when they crushed their targets.

Select the rows from sales_vs_targets where revenue is greater than target. Save these rows to the variable crushing_it.

In [12]:
crushing_it = sales_vs_targets[sales_vs_targets.revenue > sales_vs_targets.target]
crushing_it

month  revenue  target
1  February      290     270
2     March      310     300

In addition to using pd.merge, each DataFrame has its own merge method. For instance, if you wanted to merge orders with customers, you could use:

`new_df = orders.merge(customers)`

This produces the same DataFrame as if we had called pd.merge(orders, customers).

We generally use this when we are joining more than two DataFrames together because we can "chain" the commands. The following command would merge orders to customers, and then the resulting DataFrame to products:

`big_df = orders.merge(customers).merge(products)`

We have some more data from Cool T-Shirts Inc. The number of men's and women's t-shirts sold per month is in a file called men_women_sales.csv. Load this data into a DataFrame called men_women.

In [14]:
men_women = pd.read_csv('men_women_sales.csv')

Merge all three DataFrames (sales, targets, and men_women) into one big DataFrame called all_data.

In [16]:
all_data = sales.merge(targets).merge(men_women)
print(all_data)

      month  revenue  target  men  women
0   January      300     310   30     35
1  February      290     270   29     35
2     March      310     300   31     29
3     April      325     350   32     28
4       May      475     475   47     50
5      June      495     500   49     45


Cool T-Shirts Inc. thinks that they have more revenue in months where they sell more women's t-shirts.

Select the rows of all_data where:

    revenue is greater than target

AND

    women is greater than men

Save your answer to the variable results.


In [18]:
results = all_data[(all_data.revenue > all_data.target) & (all_data.women > all_data.men)]
print(results)

      month  revenue  target  men  women
1  February      290     270   29     35


## Merge on Specific Columns

In the previous example, the merge function "knew" how to combine tables based on the columns that were the same between two tables. For instance, products and orders both had a column called product_id. This won't always be true when we want to perform a merge.

Generally, the products and customers DataFrames would not have the columns product_id or customer_id. Instead, they would both be called id and it would be implied that the id was the product_id for the products table and customer_id for the customers table. They would look like this:

### Customers

<table>
    <tr><th>id</th><th>customer_name</th><th>address</th><th>phone_number</th></tr>
    <tr><th>1</th><td>John Smith</td><td>123 Main St.</td><td>212-123-4567</td></tr>
    <tr><th>2</th><td>Jane Doe</td><td>456 Park Ave.</td><td>949-867-5309</td></tr>
    <tr><th>3</th><td>Joe Schmo</td><td>798 Broadway</td><td>112-358-1321</td></tr>
</table>

### Products

<table>
    <tr><th>id</th><th>description</th><th>price</th></tr>
    <tr><td>1</td><td>thing-a-ma-jig</td><td>5</td></tr>
    <tr><td>2</td><td>whatcha-ma-call-it</td><td>10</td></tr>
    <tr><td>3</td><td>doo-hickey</td><td>7</td></tr>
    <tr><td>4</td><td>gizmo</td><td>3</td></tr>
</table>

## How would this affect our merges?

Because the id columns would mean something different in each table, our default merges would be wrong.

One way that we could address this problem is to use .rename to rename the columns for our merges. In the example below, we will rename the column id to customer_id, so that orders and customers have a common column for the merge.

`pd.merge(orders,customers.rename(columns={'id': 'customer_id'}))`

In [20]:
orders_products = pd.merge(orders,products.rename(columns={'id':'product_id'}))
print(orders_products)

   order_id  customer_id  product_id  quantity   timestamp  \
0         1            2           3         1  2017-01-01   
1         5            3           3         3  2017-02-01   
2         2            2           2         3  2017-01-01   
3         4            3           2         2  2017-02-01   
4         3            3           1         1  2017-01-01   
5         7            1           1         1  2017-02-02   
6         6            1           4         2  2017-03-01   
7         8            1           4         1  2017-02-02   

          description  price  
0          doo-hickey      7  
1          doo-hickey      7  
2  whatcha-ma-call-it     10  
3  whatcha-ma-call-it     10  
4      thing-a-ma-jig      5  
5      thing-a-ma-jig      5  
6               gizmo      3  
7               gizmo      3  


In the previous exercise, we learned how to use rename to merge two DataFrames whose columns don't match.

If we don't want to do that, we have another option. We could use the keywords left_on and right_on to specify which columns we want to perform the merge on. In the example below, the "left" table is the one that comes first (orders), and the "right" table is the one that comes second (customers). This syntax says that we should match the customer_id from orders to the id in customers.

`pd.merge(orders,customers,left_on='customer_id',right_on='id')`

If we use this syntax, we'll end up with two columns called id, one from the first table and one from the second. Pandas won't let you have two columns with the same name, so it will change them to id_x and id_y.

It will look like this: 

<table>
    <tr><th>id_x</th><th>customer_id</th><th>product_id</th><th>quantity</th><th>timestamp</th><th>id_y</th><th>customer_name</th><th>address</th><th>phone_number</th></tr>
    <tr><td>1</td><td>2</td><td>3</td><td>1</td><td>2017-01-01 00:00:00</td><td>2</td><td>Jane Doe</td><td>456 Park Ave.</td><td>949-867-5309</td></tr>
    <tr><td>2</td><td>2</td><td>2</td><td>3</td><td>2017-01-01 00:00:00</td><td>2</td><td>Jane Doe</td><td>456 Park Ave.</td><td>949-867-5309</td></tr>
    <tr><td>3</td><td>3</td><td>1</td><td>1</td><td>2017-01-01 00:00:00</td><td>3</td><td>Joe Schmo</td><td>798 Broadway</td><td>112-358-1321</td></tr>
    <tr><td>4</td><td>3</td><td>2</td><td>2</td><td>2017-02-01 00:00:00</td><td>3</td><td>Joe Schmo</td><td>798 Broadway</td><td>112-358-1321</td></tr>
    <tr><td>5</td><td>3</td><td>3</td><td>3</td><td>2017-02-01 00:00:00</td><td>3</td><td>Joe Schmo</td><td>798 Broadway</td><td>112-358-1321</td></tr>
    <tr><td>6</td><td>1</td><td>4</td><td>2</td><td>2017-03-01 00:00:00</td><td>1</td><td>John Smith</td><td>123 Main St.</td><td>212-123-4567</td></tr>
    <tr><td>7</td><td>1</td><td>1</td><td>1</td><td>2017-02-02 00:00:00</td><td>1</td><td>John Smith</td><td>123 Main St.</td><td>212-123-4567</td></tr>
    <tr><td>8</td><td>1</td><td>4</td><td>1</td><td>2017-02-02 00:00:00</td><td>1</td><td>John Smith</td><td>123 Main St.</td><td>212-123-4567</td></tr>
</table>

The new column names id_x and id_y aren't very helpful for us when we read the table. We can help make them more useful by using the keyword suffixes. We can provide a list of suffixes to use instead of "_x" and "_y".

For example, we could use the following code to make the suffixes reflect the table names:

`pd.merge(orders,customers,left_on='customer_id',right_on='id',suffixes=['_order', '_customer'])`

The resulting table would look like this: 

<table>
    <tr><th>id_order</th><th>customer_id</th><th>product_id</th><th>quantity</th><th>timestamp</th><th>id_customer</th><th>customer_name</th><th>address</th><th>phone_number</th></tr>
    <tr><td>1</td><td>2</td><td>3</td><td>1</td><td>2017-01-01 00:00:00</td><td>2</td><td>Jane Doe</td><td>456 Park Ave.</td><td>949-867-5309</td></tr>
    <tr><td>2</td><td>2</td><td>2</td><td>3</td><td>2017-01-01 00:00:00</td><td>2</td><td>Jane Doe</td><td>456 Park Ave.</td><td>949-867-5309</td></tr>
    <tr><td>3</td><td>3</td><td>1</td><td>1</td><td>2017-01-01 00:00:00</td><td>3</td><td>Joe Schmo</td><td>798 Broadway</td><td>112-358-1321</td></tr>
    <tr><td>4</td><td>3</td><td>2</td><td>2</td><td>2017-02-01 00:00:00</td><td>3</td><td>Joe Schmo</td><td>798 Broadway</td><td>112-358-1321</td></tr>
    <tr><td>5</td><td>3</td><td>3</td><td>3</td><td>2017-02-01 00:00:00</td><td>3</td><td>Joe Schmo</td><td>798 Broadway</td><td>112-358-1321</td></tr>
    <tr><td>6</td><td>1</td><td>4</td><td>2</td><td>2017-03-01 00:00:00</td><td>1</td><td>John Smith</td><td>123 Main St.</td><td>212-123-4567</td></tr>
    <tr><td>7</td><td>1</td><td>1</td><td>1</td><td>2017-02-02 00:00:00</td><td>1</td><td>John Smith</td><td>123 Main St.</td><td>212-123-4567</td></tr>
    <tr><td>8</td><td>1</td><td>4</td><td>1</td><td>2017-02-02 00:00:00</td><td>1</td><td>John Smith</td><td>123 Main St.</td><td>212-123-4567</td></tr>
</table>

Merge orders and products using left_on and right_on. Use the suffixes _orders and _products. Save your results to the variable orders_products.

In [22]:
orders_products = pd.merge(orders,products,left_on = 'product_id',right_on = 'order_id',suffixes = ['_orders', '_products'])
print(orders_products)

KeyError: 'order_id'

Outer Merge

In the previous exercise, we saw that when we merge two DataFrames whose rows don't match perfectly, we lose the unmatched rows.

This type of merge (where we only include matching rows) is called an inner merge. There are other types of merges that we can use when we want to keep information from the unmatched rows.

Suppose that two companies, Company A and Company B have just merged. They each have a list of customers, but they keep slightly different data. Company A has each customer's name and email. Company B has each customer's name and phone number. They have some customers in common, but some are different.

* company_a

<table>
    <tr><th>name</th><th>email</th></tr>
    <tr><td>Sally Sparrow</td><td>sally.sparrow@gmail.com</td></tr>
    <tr><td>Peter Grant</td><td>pgrant@yahoo.com</td></tr>
    <tr><td>Leslie May</td><td>leslie_may@gmail.com</td></tr>
</table>

* company_b

<table>
    <tr><th>name</th><th>phone</th></tr>
    <tr><td>Peter Grant</td><td>626-987-6543</td></tr>
    <tr><td>Leslie May</td><td>626-987-6453</td></tr>
    <tr><td>Aaron Burr</td><td>303-456-7891</td></tr>
</table>

If we wanted to combine the data from both companies without losing the customers who are missing from one of the tables, we could use an Outer Join. An Outer Join would include all rows from both tables, even if they don't match. Any missing values are filled in with None or nan (which stands for "Not a Number").

`pd.merge(company_a, company_b, how='outer')`

The resulting table would look like this:

<table>
    <tr><th>name</th><th>email</th><th>phone</th></tr>
    <tr><td>Sally Sparrow</td><td>sally.sparrow@gmail.com</td><td>nan</td></tr>
    <tr><td>Peter Grant</td><td>pgrant@yahoo.com</td><td>212-345-6789</td></tr>
    <tr><td>Leslie May</td><td>leslie_may@gmail.com</td><td>626-987-6543</td></tr>
    <tr><td>Aaron Burr</td><td>nan</td><td>303-456-7891</td></tr>
</table>

There are two hardware stores in town: Store A and Store B. Store A's inventory is in DataFrame store_a and Store B's inventory is in DataFrame store_b. They have decided to merge into one big Super Store!

Combine the inventories of Store A and Store B using an outer merge. Save the results to the variable store_a_b_outer.

In [25]:
store_a = pd.read_csv('store_a.csv')
store_b = pd.read_csv('store_b.csv')

store_a_b_outer = pd.merge(store_a, store_b, how='outer')
print(store_a_b_outer)

             item  store_a_inventory  store_b_inventory
0          hammer               12.0                6.0
1     screwdriver               15.0                NaN
2           nails              200.0              250.0
3          screws              350.0                NaN
4             saw                6.0                6.0
5       duct tape              150.0              150.0
6          wrench               12.0                NaN
7        pvc pipe               54.0               54.0
8            rake                NaN               10.0
9          shovel                NaN               15.0
10  wooden dowels                NaN              192.0


## Left Merge

Let's return to the merge of Company A and Company B.

Suppose we want to identify which customers are missing phone information. We would want a list of all customers who have email, but don't have phone.

We could get this by performing a Left Merge. A Left Merge includes all rows from the first (left) table, but only rows from the second (right) table that match the first table.

For this command, the order of the arguments matters. If the first DataFrame is company_a and we do a left join, we'll only end up with rows that appear in company_a.

By listing company_a first, we get all customers from Company A, and only customers from Company B who are also customers of Company B.

`pd.merge(company_a, company_b, how='left')`

The result would look like this:

<table>
    <tr><th>name</th><th>email</th><th>phone</th></tr>
    <tr><td>Sally Sparrow</td><td>sally.sparrow@gmail.com</td><td>None</td></tr>
    <tr><td>Peter Grant</td><td>pgrant@yahoo.com</td><td>212-345-6789</td></tr>
    <tr><td>Leslie May</td><td>leslie_may@gmail.com</td><td>626-987-6543</td></tr>
</table>

If we reverse the order of company_b and company_a, we'll only get rows that appear in company_b:

`pd.merge(company_b, company_a, how='left')`

<table>
    <tr><th>name</th><th>email</th><th>phone</th></tr>
    <tr><td>Peter Grant</td><td>pgrant@yahoo.com</td><td>212-345-6789</td></tr>
    <tr><td>Leslie May</td><td>leslie_may@gmail.com</td><td>626-987-6543</td></tr>
    <tr><td>Aaron Burr</td><td>None</td><td>303-456-7891</td></tr>
</table>

Let's return to the two hardware stores, Store A and Store B. They're not quite sure if they want to merge into a big Super Store just yet.

Store A wants to find out what products they carry that Store B does not carry. Using a left merge, combine store_a to store_b and save the results to store_a_b_left.

The items with null in store_b_inventory are carried by Store A, but not Store B.

In [28]:
store_a_b_left = pd.merge(store_a, store_b, how='left')
store_b_a_left = pd.merge(store_b, store_a, how='left')

print(store_a_b_left,'\n')
print(store_b_a_left)

          item  store_a_inventory  store_b_inventory
0       hammer                 12                6.0
1  screwdriver                 15                NaN
2        nails                200              250.0
3       screws                350                NaN
4          saw                  6                6.0
5    duct tape                150              150.0
6       wrench                 12                NaN
7     pvc pipe                 54               54.0 

            item  store_b_inventory  store_a_inventory
0         hammer                  6               12.0
1          nails                250              200.0
2            saw                  6                6.0
3      duct tape                150              150.0
4       pvc pipe                 54               54.0
5           rake                 10                NaN
6         shovel                 15                NaN
7  wooden dowels                192                NaN


## Concatenate DataFrames

Sometimes, a dataset is broken into multiple tables. For instance, data is often split into multiple CSV files so that each download is smaller.

When we need to reconstruct a single DataFrame from multiple smaller DataFrames, we can use the method pd.concat([df1, df2, df2, ...]). This method only works if all of the columns are the same in all of the DataFrames.

For instance, suppose that we have two DataFrames:

* df1

<table>
    <tr><th>name</th><th>email</th></tr>
    <tr><td>Katja Obinger</td><td>k.obinger@gmail.com</td></tr>
    <tr><td>Alison Hendrix</td><td>alisonH@yahoo.com</td></tr>
    <tr><td>Cosima Niehaus</td><td>cosi.niehaus@gmail.com</td></tr>
    <tr><td>Rachel Duncan</td><td>rachelduncan@hotmail.com</td></tr>
</table>

* df2

<table>
    <tr><th>name</th><th>email</th></tr>
    <tr><td>Jean Gray</td><td>jgray@netscape.net</td></tr>
    <tr><td>Scott Summers</td><td>ssummers@gmail.com</td></tr>
    <tr><td>Kitty Pryde</td><td>kitkat@gmail.com</td></tr>
    <tr><td>Charles Xavier</td><td>cxavier@hotmail.com</td></tr>
</table>

If we want to combine these two DataFrames, we can use the following command:

`pd.concat([df1, df2])`

That would result in the following DataFrame:

<table>
    <tr><th>name</th><th>email</th></tr>
    <tr><td>Katja Obinger</td><td>k.obinger@gmail.com</td></tr>
    <tr><td>Alison Hendrix</td><td>alisonH@yahoo.com</td></tr>
    <tr><td>Cosima Niehaus</td><td>cosi.niehaus@gmail.com</td></tr>
    <tr><td>Rachel Duncan</td><td>rachelduncan@hotmail.com</td></tr>
    <tr><td>Jean Gray</td><td>jgray@netscape.net</td></tr>
    <tr><td>Scott Summers</td><td>ssummers@gmail.com</td></tr>
    <tr><td>Kitty Pryde</td><td>kitkat@gmail.com</td></tr>
    <tr><td>Charles Xavier</td><td>cxavier@hotmail.com</td></tr>
</table>

An ice cream parlor and a bakery have decided to merge.

The bakery's menu is stored in the DataFrame bakery, and the ice cream parlor's menu is stored in DataFrame ice_cream.

Create their new menu by concatenating the two DataFrames into a DataFrame called menu.

In [31]:
bakery = pd.read_csv('bakery.csv')
ice_cream = pd.read_csv('ice_cream.csv')

menu = pd.concat([bakery, ice_cream])
print(menu)

                              item  price
0                           cookie   2.50
1                          brownie   3.50
2                    slice of cake   4.75
3              slice of cheesecake   4.75
4                     slice of pie   5.00
0     scoop of chocolate ice cream   3.00
1       scoop of vanilla ice cream   2.95
2    scoop of strawberry ice cream   3.05
3  scoop of cookie dough ice cream   3.25


Cool T-Shirts Inc. just created a website for ordering their products. They want you to analyze two datasets for them:

* visits contains information on all visits to their landing page
* checkouts contains all users who began to checkout on their website

Use print to inspect each DataFrame.

In [37]:
visits = pd.read_csv('visits.csv',
                        parse_dates=[1])
print(visits.head(),'\n')
checkouts = pd.read_csv('checkouts.csv',
                        parse_dates=[1])
print(checkouts.head())

                                user_id          visit_time
0  319350b4-9951-47ef-b3a7-6b252099905f 2017-02-21 07:16:00
1  7435ec9f-576d-4ebd-8791-361b128fca77 2017-05-16 08:37:00
2  0b061e73-f709-42fa-8d1a-5f68176ff154 2017-04-12 19:32:00
3  9133d6f0-e68b-4c8d-bafd-ff2825e8dafe 2017-08-18 04:32:00
4  08d13edb-071c-4cfb-9ee4-8f377d0e932a 2017-07-08 06:24:00 

                                user_id       checkout_time
0  fe90a9f4-960a-4a0d-9160-e562adb79365 2017-11-09 09:25:00
1  1a35b7eb-f603-407d-91be-a2c3304066fd 2017-08-15 21:25:00
2  e2c24ee0-7fdf-4400-abde-b36378fe5ce6 2017-07-04 15:39:00
3  10dbd3c5-d610-44e9-9994-110a7950b6b4 2017-08-09 21:07:00
4  f028e9dd-77d0-4002-83f6-372a4837fda6 2017-10-27 08:57:00


We want to know the amount of time from a user's initial visit to the website to when they start to check out.

Use merge to combine visits and checkouts and save it to the variable v_to_c.

In [38]:
v_to_c = pd.merge(visits, checkouts)

In order to calculate the time between visiting and checking out, define a column of v_to_c called time by pasting the following code into script.py:

```
v_to_c['time'] = v_to_c.checkout_time - v_to_c.visit_time
print(v_to_c)
```

In [39]:
v_to_c['time'] = v_to_c.checkout_time - v_to_c.visit_time
print(v_to_c)

                                 user_id          visit_time  \
0   319350b4-9951-47ef-b3a7-6b252099905f 2017-02-21 07:16:00   
1   319350b4-9951-47ef-b3a7-6b252099905f 2017-02-21 07:16:00   
2   7435ec9f-576d-4ebd-8791-361b128fca77 2017-05-16 08:37:00   
3   7435ec9f-576d-4ebd-8791-361b128fca77 2017-05-16 08:37:00   
4   08d13edb-071c-4cfb-9ee4-8f377d0e932a 2017-07-08 06:24:00   
5   c4dac0f2-2fa9-48a8-b056-c3b2a5a5c683 2017-07-09 14:19:00   
6   c4dac0f2-2fa9-48a8-b056-c3b2a5a5c683 2017-07-09 14:19:00   
7   c4dac0f2-2fa9-48a8-b056-c3b2a5a5c683 2017-07-09 14:19:00   
8   c4dac0f2-2fa9-48a8-b056-c3b2a5a5c683 2017-07-09 14:19:00   
9   f028e9dd-77d0-4002-83f6-372a4837fda6 2017-10-27 08:46:00   
10  746631d2-35d5-441e-a21b-e5f39442f981 2017-06-19 23:34:00   
11  746631d2-35d5-441e-a21b-e5f39442f981 2017-06-19 23:34:00   
12  a0fc94a2-4a80-4a33-994b-75783066ac62 2017-05-11 13:07:00   
13  e2c24ee0-7fdf-4400-abde-b36378fe5ce6 2017-07-04 15:33:00   
14  e6c7ecb9-4710-4cbd-ad02-c43971ebbe7f

To get the average time to checkout, paste the following code into script.py:

`print(v_to_c.time.mean())`

In [40]:
print(v_to_c.time.mean())

0 days 00:15:24.750000
